### PR TITLE
#632 : Move stack trace formatting out of Span class

### LINKED ISSUE #632: Move stack trace formatting out of Span class
The method [formatStackTrace ](https://github.com/open-telemetry/opentelemetry-php/blob/main/src/SDK/Trace/Span.php#L175) should be moved from `Span` into a dedicated Util/Formatter class in the `SDK\Common` namespace since formatting stack traces is not the responsibility of a `Span`.

- Move logic in `formatStackTrace`  from `Span` into a dedicated Util/Formatter class
- Add Backward compatibility by calling the new method from `Span::formatStackTrace` and tigger deprecation notice with [Compatibility\Util::triggerMethodDeprecationNotice](https://github.com/open-telemetry/opentelemetry-php/blob/main/src/SDK/Common/Dev/Compatibility/Util.php#L23)

### PR DESCRIPTION
**Pull request by amber0612**
closes #632

### CODE REVIEW

**Review by brettmc — APPROVED** [review]

**Comment by brettmc** [comment]
Can you sort out signing a CLA please? (We can't accept your contribution without this) And also try to write a unit test for the newly-created class, since code coverage has dropped with this refactor.

**Comment by amber0612** [comment]
> Can you sort out signing a CLA please? (We can't accept your contribution without this) And also try to write a unit test for the newly-created class, since code coverage has dropped with this refactor.

Hi Bret, Unit test for formatStackTrace was not written previously. Can u help me with what am i supposed to assert equal for Unit test of the formatStackTrace function. As the function returns String which is stack Trace and string returned will be complex with Line numbers too. Do u have ideas on how to handle unit test for this...

**Comment by Nevay** [comment]
I would write them as `.phpt` tests to have a clean stacktrace. There is an open issue with PHPUnit adding an additional frame; we can use [run-tests.php](https://github.com/php/php-src/blob/master/run-tests.php) instead (at least for now).

```php
--TEST--
Basic stacktrace format
--FILE--
<?php
use OpenTelemetry\SDK\Common\Util\TracingUtil;

require_once 'vendor/autoload.php';

echo TracingUtil::formatStackTrace(new Exception('message')), "\n";
?>
--EXPECT--
Exception: message
 at main(test_stacktrace_basic.php:6)
```

(I plan on creating a follow-up PR to resolve some issues with the current implementation and could take care of the tests if necessary.)

File: src/SDK/Common/Util/TracingUtil.php [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+use function array_key_exists;
+use function array_shift;
+use function basename;
+use function in_array;
+use function sprintf;
+use function str_replace;
+
+use Throwable;
+
+class TracingUtil
```
Comment (tidal) [inline src/SDK/Common/Util/TracingUtil.php:16]: Could you please rename the class to smth like `StackTraceFormatter` and the method to just `format`, please?

File: src/SDK/Common/Util/TracingUtil.php [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+         *  at parent_func(test.php:51)
+         *  at child_func(test.php:44)
+         *  at (main)(test.php:62)
+         *
+         * Credit: https://www.php.net/manual/en/exception.gettraceasstring.php#114980
+         *
+         */
+    public static function formatStackTrace(Throwable $e, array &$seen = null): string
+    {
+        $starter = $seen ? 'Caused by: ' : '';
```
Comment (tidal) [inline src/SDK/Common/Util/TracingUtil.php:35]: Could you make `Caused by: ` a constant, please?

File: src/SDK/Common/Util/TracingUtil.php [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+        $result = [];
+        if (!$seen) {
```
Comment (tidal) [inline src/SDK/Common/Util/TracingUtil.php:37]: Could you please check for a `null` value here, please?

File: src/SDK/Common/Util/TracingUtil.php [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+            }
+
+            // Lambda to format traces -- we want to format the trace with '.' separators
+            $slashToDot = function ($str) {
+                return str_replace('\\', '.', $str);
+            };
+
+            $traceHasKey = array_key_exists(0, $trace);
+            $traceKeyHasClass = $traceHasKey && array_key_exists('class', $trace[0]);
+            $traceKeyHasFunction = $traceKeyHasClass && array_key_exists('function', $trace[0]);
```
Comment (tidal) [inline src/SDK/Common/Util/TracingUtil.php:60]: Could you please male `class`and `function` class constants, please?

File: src/SDK/Common/Util/TracingUtil.php [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+            $result[] = sprintf(
+                ' at %s%s%s(%s%s%s)',
+                $traceKeyHasClass ? $slashToDot($trace[0]['class']) : '',
+                $traceKeyHasFunction ? '.' : '',
+                $traceKeyHasFunction ? $slashToDot($trace[0]['function']) : 'main',
+                $line === null ? $file : basename($file),
+                $line === null ? '' : ':',
+                $line ?? ''
+            );
+            $seen[] = "$file:$line";
```
Comment (tidal) [inline src/SDK/Common/Util/TracingUtil.php:71]: Could you make this a `sprintf` call, please?

File: src/SDK/Common/Util/TracingUtil.php [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+            if (!count($trace)) {
```
Comment (tidal) [inline src/SDK/Common/Util/TracingUtil.php:72]: Could you please check for a numeric value here, please?

File: src/SDK/Common/Util/TracingUtil.php [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+                break;
+            }
+            $file = array_key_exists('file', $trace[0]) ? $trace[0]['file'] : 'Unknown Source';
+            $line = array_key_exists('file', $trace[0]) && array_key_exists('line', $trace[0]) && $trace[0]['line'] ? $trace[0]['line'] : null;
```
Comment (tidal) [inline src/SDK/Common/Util/TracingUtil.php:76]: Could you make `line` a class constant, please?

File: src/SDK/Common/Util/TracingUtil.php [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+            array_shift($trace);
+        }
+        $result = implode("\n", $result);
+        if ($prev) {
```
Comment (tidal) [inline src/SDK/Common/Util/TracingUtil.php:80]: Could you check for a type here, please?

**Review by tidal — CHANGES_REQUESTED** [review]

File: src/SDK/Common/Util/TracingUtil.php (hunk 3 above)
Comment (tidal) [reply src/SDK/Common/Util/TracingUtil.php:37]: Yes

**Review by tidal — APPROVED** [review]

**Pull request merged**